nixos/thermald: add manual config generation

jakefwoods opened the pull request:
###### Motivation for this change

thermald has two modes: zero-config and manual. Sometimes it is useful to manually configure thermald to achieve better thermal results or to give thermald a hand when detecting possible cooling options.

I have tested this change on my own laptop running NixOS and observed significant improvements in the thermals of my laptop.

With this change manual thermald config can be encoded in `configuration.nix`. For example:

```
  services.thermald = {
    enable = true;
    configFile.platforms = [{
      Name = "CPU passive cooling limit to 80C";
      ProductName = "*";
      Preference = "QUIET";
      ThermalZones = [{
        Type = "x86_pkg_temp";
        TripPoints = [{
          SensorType = "x86_pkg_temp";
          Type = "passive";
          Temperature = 80000;
          CoolingDevices = [{
            Index = 1;
            Type = "rapl_controller";
            Influence = 50;
            SamplingPeriod = 10;
          } {
            Index = 2;
            Type = "intel_pstate";
            Influence = 40;
            SamplingPeriod = 10;
          } {
            Index = 3;
            Type = "intel_powerclamp";
            Influence = 30;
            SamplingPeriod = 10;
          } {
            Index = 4;
            Type = "cpufreq";
            Influence = 20;
            SamplingPeriod = 8;
          } {
            Index = 5;
            Type = "Processor";
            Influence = 10;
            SamplingPeriod = 5;
          }];
        }];
      }];
    }];
  };
```

I'm not sure if the naming is exactly what we want, particularly around `configFile.platforms` so I'd appreciate any input on how to make this more consistent with other NixOS modules.

I'd also like some feedback on how to better implement `TripPoints.CoolingDevices`. At the moment I've implemented `CoolingDevices` as a list but in the actual XML it simply allows for multiple `CoolingDevice` entries under `TripPoints`.

Currently we do:

```
TripPoints = [{
  ...
  CoolingDevices = [{
    ...
  }, {
    ...
  }];
}];
```

But if we want to mirror how the XML is actually represented it would be closer to do:

```
TripPoints = [{
  ...
  CoolingDevice = { ... };
  CoolingDevice = { ... };
}];
```

But I'm unsure how to get submodules to merge correctly given the above syntax. Any input would be appreciated \:)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Infinisil commented:
I'm not sure if it's worth adding almost 40 options for the small number of people that will actually use them. NixOS evaluation time is still about linear in the amount of options, and as long as that's not resolved I think it would be wise not to use options for every single thing. Although, probably in the long term we shouldn't refrain from adding lots of options for things. But we have no idea how long it will be until somebody comes up with a solution to the evaluation time problem (I've been thinking about it myself for a while now, I have an idea, but no implementation yet).

As an alternative, see how I use a single option to represent the whole config in  https://github.com/NixOS/nixpkgs/pull/41467, and then use this to generate a syntactically valid config file. This has the advantage of only using one option and not needing NixOS module adjustments with new versions, but the disadvantage of not being able to check for valid config settings with the module system. What still can be done is to use the package itself to check for valid config settings at build time (by running it with some config check option, if supported).

@Vodurden What do you think about this? I'm fine with either adding all the options or doing it the alternate way. I'll start reviewing the code after this is cleared up.

Vodurden commented:
@Infinisil: Ooh that's an interesting approach, if I understand it correctly you're building a small abstract syntax tree using nix's semantic types and using it to represent the configuration file which you then compile into the config. 

I'll give it a shot and see how the resulting code turns out. Though it is a shame not to be able to use the options system to it's fullest as I _think_ we lose some of the documentation discoverability which `thermald` definitely needs.

Vodurden commented:
I just had a quick look into this: `thermald` doesn't seem to have a config file verification mode so I won't be able to use exactly the same approach and maintain config file checking. 

With that in mind I have a few options:

### Use nixos options

Pros:

- Current state, no extra work for me
- Full documentation for all options
- Static checking for config

Cons:

- Will negatively impact nixos evaluation time due to a bug that may take a long time to be fixed.

### Use a generic semantic type and compile it to config

Pros:

- No impact on nixos evaluation time
- Doesn't require updates to the module if new config is added to `thermald`

Cons:

- Config cannot be checked for validity unless I do something like writing a DTD for thermald. (`thermald` doesn't have a config checker).
- No documentation for config options 

### Use a specific semantic type and compile it to config

This option is the same as the above except I could encode a more specific type system that matches the allowable config in `thermald`. (I haven't tried this option so I'm not sure if it will actually work but it feels theoretically possible)

Pros:

- No impact on nixos evaluation time
- Config can be checked for validity

Cons:

- No documentation for config options
- Requires updates if `thermald` config changes


## What to do?

Overall I'm not sure what approach to take. I'm not a big fan of introducing a generic config syntax tree since it won't have any static checking and will fail at runtime. But I also don't want to slow down the evaluation time for everything.

Ultimately I guess what I need to know is:

- Is the slowdown to nixos evaluation significant enough to be worth doing a bunch of work?
- Would we be happy with the generic semantic type config approach even if the results can't be validated?
- Would we be happy with the specific semantic type config approach if I can make it work?

Infinisil commented:
Neat overview! So I now think that doing it the current way is probably better. The point that convinced me is that docs for the available options are really nice, especially because thermald's config file docs are kinda bad. The NixOS module can serve as a better replacement for them.

Very nice first PR btw! I'll try to review the code today :)

Infinisil commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -4,27 +4,392 @@ with lib;
+      (xmlElement cdCfg.PidControl "ki")
+    ])
+    (xmlElement cdCfg "WritePrefix")
+  ];
```
I think a generic xml generator would be better than this. Something like

```nix
{
  config = {
    ThermalConfiguration.Platform = map (platform: {
      Name = platform.name;
      UUID = platform.UUID;
      # ...
      TermalSensors.ThermalSensor = map (sensor: {
        Name = sensor.name;
        # ...
      }) platform.thermalSensors;
      # ...
    }) configFile.platforms;
  };
  
  toXml = val:
    # Generate xml value by using the type of `val`:
    # toString for simple values
    # for attrsets use name/value as xml attr and value (recursively). if value is a list, generate multiple with the same name
  ;

  configXml = "<?xml version..." + concatStrings (toXml cfg.config);
}
```

You might want to make `toXML` return a list of strings, representing lines of the config file, then you can easily indent them when you recurse by mapping over them.

I did that for my ZNC PR (its syntax is not xml unfortunately): https://github.com/NixOS/nixpkgs/pull/41467/files#diff-f93a99c1b50dc7cd526921714cb77de1R22

Vodurden replied:
I had a go at using a generic representation but I've hit a snag that I'm not sure how to resolve.

Currently I've written `toXML` like so:

```
  toXml =
    let listToXml = list: concatLines (map toXml' list);
        attrToXml = key: value:
          if skipAttribute value
          then ""
          else "<${key}>" + (indentIfNeeded value (toXml' value)) + "</${key}>";
        setToXml = set: (concatLines (map (k: attrToXml k (set.${k})) (builtins.attrNames set)));

        # We don't want to include empty lines in the output as it bloats the XML for no reason.
        concatLines = lines: builtins.concatStringsSep "\n" (lib.remove "" lines);

        # We want to completel remove some attributes based on their value as they provide no value
        # to the output and cause bloating if left in.
        skipAttribute = value: value == null || value == [];

        # We don't want to indent strings since they're usually short one-liners
        skipIndent = value: (builtins.typeOf value == "string");
        indentIfNeeded = value: xml:
          if skipIndent value then xml else "\n" + (indent xml) + "\n";

        toXml' = value:
                if builtins.typeOf value == "list" then listToXml value
           else if builtins.typeOf value == "set"  then setToXml value
           else builtins.toString value;
    in toXml'
```

This works fine for the mots of the configuration but there's a particular configuration detail that I don't know how to represent using sets.

When defining multiple `CoolingDevice`s in a trip point the XML expects the following format:

```
<TripPoint>
  ...
  <CoolingDevice>
    ...
  </CoolingDevice>
  <CoolingDevice>
    ...
  </CoolingDevice>
</TripPoint>
```

This is different to every other list in the config file in that it _doesn't_ have a parent `<CoolingDevices>` tag. As a result I would need some way to indicate in a set that multiple keys should be merged inline. I'd imagine it would look something like this:

```
tripPoint: {
  CoolingDevice = ...
  CoolingDevice = ...
}
```

But as far as I understand it, sets cannot have duplicate entries.

In the current solution I solve it by introducing a `CoolingDevices` attribute and then interpreting that as a special case:

```
  makeTripPointXml = tpCfg: xmlBlock "TripPoint" ([
    (xmlElement tpCfg "SensorType")
    (mapXmlElement toString tpCfg "Temperature")
    (xmlElement tpCfg "Type")
    (xmlElement tpCfg "ControlType")
  ] ++ (map makeTripPointCoolingDeviceXml tpCfg.CoolingDevices));
```

However I don't believe I can do this with a generic builder.

Please let me know if there's a way around this, otherwise I think it might make sense to stick with the `xmlElement` and `xmlBlock` based approach.

Infinisil commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -4,27 +4,392 @@ with lib;
+      };
+
+      debug = mkOption {
+        default = false;
```
Missing type

Infinisil commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -4,27 +4,392 @@ with lib;
+          options = {
+            platforms = mkOption {
+              default = [];
+              type = with types; listOf platformConfig;
```
Nit: `with types; listOf platformConfig` -> `types.listOf platformConfig`

Infinisil commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -4,27 +4,392 @@ with lib;
+      script = ''
+        exec ${pkgs.thermald}/sbin/thermald \
+          --no-daemon \
+          ${optionalString cfg.debug "--loglevel=debug"} \
```
This can be put into `serviceConfig.ExecStart` directly, without the `exec`

Infinisil commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -4,27 +4,392 @@ with lib;
+        '';
+      };
+
+      configFile = mkOption {
```
I think it would be better to name this option `config` and introduce a new option `configFile`, which will default to the generated config, but will then be overridable by the user. Then `cfg.configFile` can also be passed directly to the exec script and to `etc`.

Infinisil commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -4,27 +4,392 @@ with lib;
+
+  platformConfig = with types; submodule {
+    options = {
+      Name = mkOption {
```
Options are named according to camelCase, so all of the first letters should be lower cased instead (except UUID).

Vodurden replied:
Would we keep UUID as is? Or should it be `uuid`?

Infinisil replied:
Oh actually a couple options seem to be using `uuid`, so that's probably better

Infinisil commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -4,27 +4,392 @@ with lib;
+      Name = mkOption {
+        type = string;
+        description = "The name of this platform";
+      };
```
A lot of options could use example values

Infinisil commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -4,27 +4,392 @@ with lib;
+
+            ki = mkOption {
+              type = string;
+              description = "ki";
```
Also a lot of options could use better descriptions, e.g. these ones. But it might be pretty hard to find out what these exactly represent.

Vodurden replied:
I'd love to give these better descriptions but I have no idea what they do. The documentation is pretty sparse on them and I'm assuming if you want to set them then you know what they are :eyes:

With that said if anyone knows what they do I'm happy to include a better description!

Vodurden commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -4,27 +4,392 @@ with lib;
+        default = null;
+        description = ''
+          Influence will be used order cooling devices. First cooling device will be
+          used, which has highest influence.
```
From what I understand this isn't _just_ an ordering, it's also used as a weight/ratio. I believe the idea is that the relative influence weight of a cooling option affects how aggressively it's used.

Maybe I can improve the description to capture this 🤔

Vodurden commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -4,27 +4,392 @@ with lib;
+        type = nullOr int;
+        default = null;
+        description = ''
+          Delay in using this cdev, this takes some time to actually cool a zone
```
I believe it's seconds but I'm not 100% sure. I'll find out and update the description \:)

Vodurden commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -4,27 +4,392 @@ with lib;
+      Type = mkOption {
+        type = string;
+        description = ''
+          The type of cooling device.
```
It's a little bit complicated. From my understanding we have:

- `ThermalZones.Type`: The type of a thermal zone, this is basically the "name" of this zone and I'm not sure if it actually refers to any system capabilities.
- `TripPoint.SensorType`: The sensor used to detect the temperature of this thermal zone. This refers to a sensor under `/sys/class/thermal` so an example could be `x86_pkg_temp`. This can also refer to a sensor defined in the `<ThermalSensors>` section.
- `CoolingDevices.Type`: The identifier of the driver to use to cool this thermal zone. Some of these can be found under `/sys/class/thermal` but some of them seem to be kernel drivers that do not appear under that path. Examples include: `rapl_controller`, `intel_pstate`, `intel_powerclamp`, `cpufreq` and `Processor`

Vodurden commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -4,27 +4,392 @@ with lib;
+      Index = mkOption {
+        type = nullOr int;
+        default = null;
+        description = "The index of this cooling device";
```
I _think_ this might be related to the index of the acpi device and so I'm not sure if we should infer it this way. 

In particular I'm thinking of it from the perspective of someone copying a config from the internet, I'd imagine they'd expect an `index` parameter and would be surprised to find it missing, so I feel like it might be worth keeping even though we _might_ be able to infer it.

What do you think?

Infinisil replied:
Hmm yeah, I'd still prefer not having unnecessary options, uniformity over NixOS options seems more important than keeping the original config file options.

Would be good to have more information on whether this index server another purpose than just telling the index in the list.

Vodurden commented:
Thanks for doing such a detailed review! I've addressed the documentation issues as best I can. I'll follow up with some more changes in the next few days to address the XML generation and configuration naming issues.

edolstra commented:
I'm reluctant about this. @Infinisil already brought up the evaluation performance issue: the NixOS evaluation model is currently not very scalable, so adding dozens of service-specific options should be avoided. 

The other argument is (lack of) maintainability: we're basically replicating the entire thermald configuration language in NixOS, so every time thermald adds an option, we'd have to update the NixOS module to make it available to users. We also have to duplicate all option descriptions from the upstream documentation, which is also not very nice - it's better to just refer to the upstream docs.

jakefwoods committed:
nixos/thermald: add manual config file

thermald has two modes: zero-config and manual. Sometimes it is useful
to manually configure thermald to achieve better thermal results or to give
thermald a hand when detecting possible cooling options.

Vodurden commented:
Ok! So in the interest of getting this merged I have removed the nixos option based configuration in favor of the proposed `configFile` attribute. This allows configuration like so:

```
  services.thermald = {
    enable = true;
    configFile = ''
      <Platforms>
        <Name>Keep CPU below 80C</Name>
        <ProductName>*</ProductName>
        <Preference>quiet</Preference>

        <ThermalZones>
          <Type>x86_pkg_temp</Type>
          <TripPoints>
            <SensorType>x86_pkg_temp</SensorType>
            <Type>passive</Type>
            <Temperature>80000</Temperature>

            <CoolingDevice>
              <Index>1</Index>
              <Type>rapl_controller</Type>
              <Influence>50</Influence>
              <SamplingPeriod>10</SamplingPeriod>
            </CoolingDevice>

            <CoolingDevice>
              <Index>2</Index>
              <Type>intel_pstate</Type>
              <Influence>40</Influence>
              <SamplingPeriod>10</SamplingPeriod>
            </CoolingDevice>

            <CoolingDevice>
              <Index>3</Index>
              <Type>intel_powerclamp</Type>
              <Influence>30</Influence>
              <SamplingPeriod>10</SamplingPeriod>
            </CoolingDevice>

            <CoolingDevice>
              <Index>4</Index>
              <Type>cpufreq</Type>
              <Influence>20</Influence>
              <SamplingPeriod>8</SamplingPeriod>
            </CoolingDevice>

            <CoolingDevice>
              <Index>5</Index>
              <Type>Processor</Type>
              <Influence>10</Influence>
              <SamplingPeriod>5</SamplingPeriod>
            </CoolingDevice>
          </TripPoints>
        </ThermalZones>
      </Platforms>
    '';
  };
```

I have separated the option based configuration into a separate commit available here: https://github.com/Vodurden/nixpkgs/commit/22d54042d026e4edee941750cef2eb0949a2efe3

If we decide to merge the follow-up commit it would be backwards compatible with this change as it simply sets the default of `configFile` to the generated config as suggested by @Infinisil.

My feeling is that _hopefully_ it's not too controversial to merge the `configFile` based approach, if there is a great desire for the option based approach I'm happy to open a follow-up PR that introduces it \:)

@Infinisil, @edolstra: Please let me know if there are any other concerns/blockers for getting this change merged and I'll do my best to resolve them as soon as I can!

infinisil reviewed (changes requested):
Yeah I think this is in the right direction, it's very common to have such an option for services. It can still be expanded to more powerful options later.

Vodurden commented:
@Infinisil: Using the `path` type makes perfect sense, I've updated this PR with the requested changes. Thanks!

Let me know if there's anything else blocking this PR from being merged

infinisil reviewed (approved):
LGTM!
